src/beliefpropagation.jl

```diff
@@ -109,27 +137,45 @@ function get_environment(tn::ITensorNetwork, mts::DataGraph, verts::Vector; dir=
     return get_environment(tn, mts, setdiff(vertices(tn), verts))
   end
 
-  env_tensors = ITensor[mts[e] for e in boundary_edges(mts, subgraphs; dir=:in)]
-  return vcat(
-    env_tensors,
-    ITensor[tn[v] for v in setdiff(flatten([vertices(mts[s]) for s in subgraphs]), verts)],
-  )
+  env_tns = ITensorNetwork[mts[e] for e in boundary_edges(mts, subgraphs; dir=:in)]
+  central_tn = ITensorNetwork([
+    tn[v] for v in setdiff(flatten([vertices(mts[s]) for s in subgraphs]), verts)
+  ])
+  return vcat(env_tns, ITensorNetwork[central_tn])
+end
+
+function get_environment(
+  output_type::Type, tn::ITensorNetwork, mts::DataGraph, verts::Vector; kwargs...
+)
+  itns = get_environment(tn::ITensorNetwork, mts::DataGraph, verts::Vector; kwargs...)
+
+  if output_type == Vector{ITensorNetwork}
+    return itns
+  else
+    itn = reduce(⊗, itns)
+    if output_type == ITensorNetwork
+      return itn
+    elseif output_type == Vector{ITensor}
+      return ITensor[itn[v] for v in vertices(itn)]
+    else
+      error("Output Type for get_environment not Supported!")
+    end
+  end
 end
```

Comment:
This is a bit confusing and hard to follow, could you explain what this is doing?

Also, it is preferable for code organization to dispatch on types instead of using if-statements like this, for example:
```julia
function get_environment(
  output_type::Type, tn::ITensorNetwork, mts::DataGraph, verts::Vector; kwargs...
)
  itns = get_environment(tn::ITensorNetwork, mts::DataGraph, verts::Vector; kwargs...)
  return _convert_itensornetworks(output_type, itns)
end

_convert_itensornetworks(::Type{<:Vector{<:ITensorNetwork}}, itns) = itns

_convert_itensornetworks(::Type{<:ITensorNetwork}, itns) = reduce(⊗, itns)

function _convert_itensornetworks(::Type{Vector{ITensor}}, itns)
  itn = _convert_itensornetworks(ITensorNetwork, itns)
  return ITensor[itn[v] for v in vertices(itn)]
end
```
though it doesn't seem like this requires a devoted `get_environment` interface to do all of this (it seems like someone could just do the conversions themselves if they want to, as a post-processing of the output of `get_environment`).

Reply:
Say we define:
```julia
ITensorNetwork(itns::Vector{ITensorNetwork}) = reduce(⊗, itns)
```
which seems reasonable to have anyway.

In that case, it doesn't seem any harder to write:
```julia
envs = ITensorNetwork(get_environment(tn, mts, verts))
```
vs.
```julia
envs = get_environment(ITensorNetwork, tn, mts, verts)
```
so it's not clear to me that conversion functionality should be in the `get_environment` function.

For now let's put aside the issue of outputting as `Vector{ITensor}`, which is already easy to convert to from an `ITensorNetwork` with:
```julia
tn = ITensorNetwork(named_grid((2, 2)); link_space=2)
ts = Vector{ITensor}(tn)
```
and I don't want to encourage people to use `Vector{ITensor}` in general anyway.

Part of my reason for trying to keep `get_environment` simpler is anticipating how that function will work for other contraction backends beyond BP, like Linjian's algorithm or CTMRG. The more features we add to it, the harder it is to design it to support various backends going forward.

Some other related questions:

1. When would you actually want `Vector{ITensorNetwork}` output instead of just an `ITensorNetwork`?
2. If there is a case where `Vector{ITensorNetwork}` is useful, is it better to output it as a partitioned graph in the standard format we use for that, i.e. as a `DataGraph`? I guess for BP they would be disconnected graphs anyway so it doesn't matter much, but that also points to the fact that in BP the format `Vector{ITensorNetwork}` doesn't seem so useful anyway, since we could just output `ITensorNetwork` and users could partition it into disconnected components if they want to distinguish between message tensor networks.

So what I'm getting it is ultimately if we can simplify all of this just by outputting an `ITensorNetwork`.

Reply:
Right, and also I think that is a case where `full_update` could just accept both as inputs and do the conversion itself.

Reply:
Okay let me change that in `full_update`